### ISSUE TITLE
Support "allOf" + "if" combination

### ISSUE THREAD

**Issue by lfyuomr-gylo** [eclipsesource/jsonforms#1518]
### Is your feature request related to a problem? Please describe.
Suppose I want to create a form for either a Point(which is denoted by coordinates) or a Line (which is denoted by coordinates of two points it's drawn through).

To cope with this,  "Understanding JSON Schema" guide [suggests](https://json-schema.org/understanding-json-schema/reference/conditionals.html) to use combination of ``allOf`` and ``if``/``then`` keywords. So here is the solution.

```json
{
  "type": "object",
  "required": ["shape"],
  "properties": {
    "shape": {
      "type": "string",
      "enum": [
        "point",
        "line"
      ]
    }
  },
  "allOf": [
    {
      "if": {"properties": {"shape": {"const": "point"}}},
      "then": {
        "properties": {
          "x": {"type": "number"},
          "y": {"type": "number"}
        }
      }
    },
    {
      "if": {"properties": {"shape": {"const": "line"}}},
      "then": {
        "properties": {
          "x1": {"type": "number"},
          "x2": {"type": "number"},
          "y1": {"type": "number"},
          "y2": {"type": "number"}
        }
      }
    }
  ]
}
```

That's what I want to have in my JSON Schema because I want to validate the structure on the backend side too. Unfortunately, currently (I'm using jsonforms ``2.3.1``) the ``if``/``then`` schema feature seems to be completely unsupported by jsonforms. 

### Describe alternatives you've considered

Of course, I can simplify the schema to a simple ``object`` with ``properties`` and move all the logic to ``rules`` block in UI Schema. However, I wouldn't be able to validate the resulting JSON on the backend side, which is a problem in my case.

### Describe the solution you'd like
It would be fantastic if I had an opportunity to attach a rule effect to a condition declared in JSON Schema, so that for aforementioned schema I would write
```json
{
  "type": "HorizontalLayout",
  "elements": [
    {
      "type": "Control",
      "scope": "#/properties/shape"
    },
    {
      "type": "Control",
      "scope": "#/properties/x",
      "rule": {
        "effect": "HIDE",
        "condition": "#/properties/allOf/0/if"
      }
    }
  ]
}
```

Unfortunately, this solution seems to be broken in case if we had the same property defined in two different ``oneOf`` branches with non-overlapping ``if`` conditions. Maybe, in this case ``$ref`` syntax would help:

```json
"rule": {
  "effect": "HIDE",
  "condition": {
    "allOf": [
      { "$ref": "#/properties/allOf/0/if" },
      { "$ref": "#/properties/allOf/5/if" }
    ]
  }
}
```

### Additional context

This issue looks very close to #595 .

**Comment by sdirix** [comment]
Thanks for your extensive research and writeup!

JSON Forms does not yet support `if`, `then`, `else` or `not` constructs. At some point we would like to support them, however it is yet undetermined when this will be the case and how the resulting UI should look like.

The good news:
In our experience `if`, `then`, `else` constructs can often just represented by the better supported `oneOf`, `allOf` and `anyOf`.

For example your schema could be refactored to this:
```
{
  "$defs": {
    "point": {
      "title": "Point",
      "properties": {
        "shape": {
          "const": "point"
        },
        "x": {
          "type": "number"
        },
        "y": {
          "type": "number"
        }
      },
      "required": ["shape"]
    },
    "line": {
      "title": "Line",
      "properties": {
        "shape": {
          "const": "line"
        },
        "x1": {
          "type": "number"
        },
        "x2": {
          "type": "number"
        },
        "y1": {
          "type": "number"
        },
        "y2": {
          "type": "number"
        }
      },
      "required": ["shape"]
    }
  },
  "oneOf": [{ "$ref": "#/$defs/point" }, { "$ref": "#/$defs/line" }]
}
```

When creating a control which points to this schema, JSON Forms will render the following:

![oneOf](https://user-images.githubusercontent.com/8998368/68837335-a5975c80-06bc-11ea-8a9b-f5ac5841822c.gif)

which is a good UI for this case in my opinion.

Drawback: At the moment the `shape` property is just ignored because it is `const`. JSON Forms will not even show a validation error as there is no associated Control for this property. We should probably support this case in some way in the future, for example by automatically adding the const value to the data.

For now you could overwrite the `oneOf` renderer and manually add the `shape` property to your data.

Please let me know if this was helpful to you.

**Comment by sdirix** [comment]
I actually thought about a different solution reusing the `if`, `then`, `else` schema. While JSON Forms has no special handling for that you can manually add your required behavior via our [rule support](https://jsonforms.io/docs/uischema/rules) in the ui schemas.

So with your original schema you can configure the ui schema like this:
```

  {
    "type": "VerticalLayout",
    "elements": [
      {
        "type": "Control",
        "scope": "#/properties/shape"
      },
      {
        "type": "Control",
        "scope": "#/allOf[0]/then/properties/x",
        "rule": {
          "effect": "SHOW",
          "condition": {
            "scope": "#/properties/shape",
            "schema": { "const": "point"}
          }
        }
      },
      {
        "type": "Control",
        "scope": "#/allOf[0]/then/properties/y",
        "rule": {
          "effect": "SHOW",
          "condition": {
            "scope": "#/properties/shape",
            "schema": { "const": "point"}
          }
        }
      },
      {
        "type": "Control",
        "scope": "#/allOf[1]/then/properties/x1",
        "rule": {
          "effect": "SHOW",
          "condition": {
            "scope": "#/properties/shape",
            "schema": { "const": "line"}
          }
        }
      },
      {
        "type": "Control",
        "scope": "#/allOf[1]/then/properties/x2",
        "rule": {
          "effect": "SHOW",
          "condition": {
            "scope": "#/properties/shape",
            "schema": { "const": "line"}
          }
        }
      },
      {
        "type": "Control",
        "scope": "#/allOf[1]/then/properties/y1",
        "rule": {
          "effect": "SHOW",
          "condition": {
            "scope": "#/properties/shape",
            "schema": { "const": "line"}
          }
        }
      },
      {
        "type": "Control",
        "scope": "#/allOf[1]/then/properties/y2",
        "rule": {
          "effect": "SHOW",
          "condition": {
            "scope": "#/properties/shape",
            "schema": { "const": "line"}
          }
        }
      }
    ]
  }
```

The resulting UI then looks like this:

![rules](https://user-images.githubusercontent.com/8998368/68851136-97096f00-06d5-11ea-9ad8-8e538a7bc8c3.gif)

This looks fine however the properties are then a little bit misplaced in the data object. You could either live with that or once again use custom controls to "fix" the path before editing the data.

**Comment by sdirix** [comment]
This issue will be closed as it's out of scope for JSON Forms at the moment. Feel free to reopen this issue if needed.

**Comment by ksze** [comment]
@sdirix If we implement our own custom renderer, like you suggest in your comment (https://github.com/eclipsesource/jsonforms/issues/1518#issuecomment-553772846), would it be possible to implement the custom render such that it can correctly "rehydrate" from existing data? Let's assume the "shape" const can be used to uniquely identify the schema among the possibilities in the `oneOf`.

**Comment by sdirix** [comment]
Hi @ksze, with "rehydrate" I assume you mean that the correct UI is shown for the corresponding `oneOf` element? If yes then that is definitely possible. In fact we already have this implemented generically in our `MaterialOneOfRenderer` by checking which of the `oneOf`-schemas fits to the existing data and then using the first one that fits.

In the case above where the data is mutually exclusive and can't overlap, i.e. conflicting `shape` values, this should always result in the correct UI to be chosen. The problem described above is that the `shape` value is neither created nor updated when switching between the variants in the `MaterialOneOfRenderer`.

The initial creation could be handled via the `default` value mechanism, however still the updating of the value has to be implemented.

So in short:
 - Yes, deciding upon the "correct" `oneOf` to choose can be done and is already implemented.
 - Creating the inital "id" (here `shape) property must be taken care of. You could use `default` for that or take care of this outside of JSON Forms or in a custom `oneOf` renderer.
 - Updating the "id" value must be implemented in a custom `oneOf` renderer.

This is useful functionality in general, so I would also be interested in a contribution which enhances the `MaterialOneOfRenderer` to do this out of the box: If all `oneOf` elements provide a similar named `const` property then treat this as an "id" element to update when switching between the variants. If there are multiple such `const` elements we could let the user configure the correct one via the ui schema options.
